secret-rotator/tests/mockclient.go

```diff
@@ -0,0 +1,213 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+// package client implements testing clients, mocked clients, and fixtures utilities.
+// Should be used with caution. Only for testing purpose.
+
+import (
+	"fmt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"strconv"
+	"time"
+)
+
+type MockClient struct {
+	Secrets map[string]map[string]*Secret
+	// Note that a new version added through MockClient.UpsertSecret() will just have
+	// a zero value of CreateTime. CreateTime generation is not mocked.
+}
+
+type Secret struct {
+	Versions map[string]*Version
+	Labels   map[string]string
+}
+
+type Version struct {
+	CreateTime time.Time
+	Data       []byte
+	State      int32
+}
+
+// ValidateProject returns nil if the project exists, otherwise error.
+func (cl *MockClient) ValidateProject(project string) error {
+	_, ok := cl.Secrets[project]
+	if !ok {
+		return status.Error(codes.NotFound, fmt.Sprintf("Project [projects/%s] not found.", project))
+	}
+
+	return nil
+}
+
+// ValidateSecret returns nil if the secret exists, otherwise error.
+func (cl *MockClient) ValidateSecret(project, id string) error {
+	_, ok := cl.Secrets[project][id]
+	if !ok {
+		return status.Error(codes.NotFound, fmt.Sprintf("Secret [projects/%s/secrets/%s] not found or has no versions.", project, id))
+	}
+
+	return nil
+}
+
+// ValidateSecretVersion returns nil if the secret version exists, otherwise error.
+func (cl *MockClient) ValidateSecretVersion(project, id, version string) error {
+	err := cl.ValidateSecret(project, id)
+	if err != nil {
+		return err
+	}
+
+	if version == "latest" {
+		version = strconv.Itoa(len(cl.Secrets[project][id].Versions))
+	}
+
+	_, ok := cl.Secrets[project][id].Versions[version]
+	if !ok {
+		return status.Error(codes.NotFound, fmt.Sprintf("Secret Version [projects/%s/secrets/%s/versions/%s] not found.", project, id, version))
+	}
+
+	return nil
+}
+
+// UpsertSecret adds a new version to the secret specified by project, id.
+// It inserts a new secret if id doesn't already exist.
+// If successful the latest version will have 'data' as its secret value,
+// and return the latest version number if successful, otherwise returns error
+func (cl *MockClient) UpsertSecret(project, id string, data []byte) (string, error) {
+	err := cl.ValidateProject(project)
+	if err != nil {
+		return "", err
+	}
+
+	err = cl.ValidateSecret(project, id)
+	if err != nil {
+		cl.Secrets[project][id] = new(Secret)
+	}
+
+	version := strconv.Itoa(len(cl.Secrets[project][id].Versions) + 1)
+	cl.Secrets[project][id].Versions[version] = &Version{
+		Data:  data,
+		State: 1, // SecretVersion_ENABLED
```

Comment:
Really feels like these should be constants somewhere.  If upstream isn't providing them, then you should define them someplace

Reply:
Switched to using `secretmanagerpb.SecretVersion_State`